### PR TITLE
[WIP][Feature] fileview page

### PR DESCRIPTION
**Pull request by chrisseto**
No description yet.

### CODE REVIEW

File: website/static/js/fangorn.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ var $ = require('jquery');
 var m = require('mithril');
 var Treebeard = require('treebeard');
 var waterbutler = require('waterbutler');
+var URI = require('../vendor/bower_components/uri.js/src/URI.js');
```
Comment (icereval) [inline website/static/js/fangorn.js]: add uri.js to webpack.common.conf.js

File: website/addons/github/static/githubFangornConfig.js [hunk 2]
```diff
@@ -5,6 +5,7 @@ var m = require('mithril');
 
 var Fangorn = require('fangorn');
 var waterbutler = require('waterbutler');
+var URI = require('../../../static/vendor/bower_components/uri.js/src/URI.js');
```
Comment (icereval) [inline website/addons/github/static/githubFangornConfig.js]: add uri.js to webpack.common.conf.js

File: website/addons/base/__init__.py [hunk 3]
```diff
@@ -166,17 +174,107 @@ class GuidFile(GuidStoredObject):
     }
 
     @property
-    def file_url(self):
+    def provider(self):
         raise NotImplementedError
 
     @property
+    def waterbutler_extras(self):
```
Comment (icereval) [inline website/addons/base/__init__.py]: Where does this ever get populated?

File: website/addons/base/views.py [hunk 4]
```diff
@@ -217,3 +222,73 @@ def get_waterbutler_render_url(**kwargs):
...
+            file_guid.enrich()
+
+        return codecs.open(file_guid.cache_path, 'r', 'utf-8').read()
+    except exceptions.AddonEnrichmentError as error:
+        return error.renderable_error
+    except IOError:
+        if start_render:
+            # Start rendering job if requested
+            download_url = furl.furl(file_guid.download_url)
+            download_url.args.update(extras)
```
Comment (icereval) [inline website/addons/base/views.py]: I'm assuming the **extras contains any addon specific request information such as specific versions, correct?

File: framework/render/tasks.py [hunk 5]
```diff
@@ -39,36 +40,42 @@ def render_mfr_error(err):
...
     :param str file_path: Full path to raw file on disk
     :param str cache_dir: Folder to store cached file in
     :param str cache_file_name: Name of cached file
     :param str download_url: External download URL
     """
-    with codecs.open(file_path) as file_pointer:
+    try:
+        ensure_path(os.path.split(temp_path)[0])
+        urllib.urlretrieve(download_url, temp_path)
+    except:
```
Comment (sloria) [inline framework/render/tasks.py]: Only catch specific exception(s). Also, does `ensure_path` need to be in the `try` block?

File: framework/render/tasks.py [hunk 6]
```diff
@@ -39,36 +40,42 @@ def render_mfr_error(err):
         """.format(**locals())
 
 
+# TODO only allow one task at a time
 @app.task(ignore_result=True, timeout=settings.MFR_TIMEOUT)
-def _build_rendered_html(file_path, cache_dir, cache_file_name, download_url):
+def _build_rendered_html(download_url, cache_path, temp_path):
     """
     :param str file_path: Full path to raw file on disk
```
Comment (sloria) [inline framework/render/tasks.py]: Please update the docstring to reflect the signature changes.

File: website/addons/base/__init__.py [hunk 7]
```diff
@@ -166,17 +174,107 @@ class GuidFile(GuidStoredObject):
...
+    @property
+    def guid_url(self):
+        return '/{0}/'.format(self._id)
+
+    @property
+    def file_name(self):
+        return '{0}.html'.format(self._id)
+
+    @property
+    def joinable_path(self):
```
Comment (sloria) [inline website/addons/base/__init__.py]: Could this just be return `self.path.lstrip('/')`?

File: website/addons/base/__init__.py [hunk 8]
```diff
@@ -166,17 +174,107 @@ class GuidFile(GuidStoredObject):
...
+        if should_raise:
+            if resp.status_code != 200:
+                raise exceptions.AddonEnrichmentError(resp.status_code)
+
+        return resp
+
+    def enrich(self, save=True):
+        resp = self._fetch_metadata(should_raise=True)
+
+        if resp.json().get('code', 200) != 200:
```
Comment (sloria) [inline website/addons/base/__init__.py]: Why is this necessary if `should_raise` is `True` in `self._fetch_metadata`?

File: framework/render/tasks.py [hunk 9]
```diff
@@ -39,8 +40,45 @@ def render_mfr_error(err):
...
+                rendered = _build_html(render_result)
+
+            # Cache rendered content
+            write_file_pointer.write(rendered)
+
+    os.remove(temp_path)
+
 @app.task(ignore_result=True, timeout=settings.MFR_TIMEOUT)
-def _build_rendered_html(file_path, cache_dir, cache_file_name, download_url):
+def _old_build_rendered_html(file_path, cache_dir, cache_file_name, download_url):
```
Comment (jmcarp) [inline framework/render/tasks.py]: Could you document the difference between `_build_rendered_html` and `_old_build_rendered_html` and explain why we have both? Do we need both?

File: website/addons/base/views.py [hunk 10]
```diff
@@ -201,19 +207,79 @@ def create_waterbutler_log(payload, **kwargs):
...
+    if not path or not node_addon:
         raise HTTPError(httplib.BAD_REQUEST)
 
-    try:
-        url = node_addon.get_waterbutler_render_url(**request.args.to_dict())
-    except TypeError:
+    if not path.startswith('/'):
+        path = '/' + path
+
+    file_guid, created = node_addon.find_or_create_file_guid(path)
```
Comment (jmcarp) [inline website/addons/base/views.py]: Does this mean we provision guids on download as well as on view?

File: website/addons/base/__init__.py [hunk 11]
```diff
@@ -166,17 +171,124 @@ class GuidFile(GuidStoredObject):
...
+    def revision(self):
+        return getattr(self, '_revision', None)
+
+    def maybe_version(self, **kwargs):
+        self._revision = kwargs.get(self.version_identifier)
+
+    def enrich(self, save=True):
+        self._fetch_metadata(should_raise=True)
+
+        if self._metadata_cache.get('code', 200) != 200:
```
Comment (jmcarp) [inline website/addons/base/__init__.py]: If WaterButler is behaving correctly, we shouldn't need this check.

File: website/addons/base/views.py (hunk 10 above)
Comment (jmcarp) [inline website/addons/base/views.py]: Would be helpful to add a base WaterButler addon class that defines `find_or_create_file_guid` as abstract.

File: website/addons/dropbox/routes.py [hunk 12]
```diff
@@ -53,31 +52,7 @@
...
-            [
-                '/project/<pid>/dropbox/files/<path:path>/download/',
-                '/project/<pid>/node/<nid>/dropbox/files/<path:path>/download/',
-            ],
-            'get',
-            views.crud.dropbox_download,
-            notemplate,
-        ),
-    ],
+    'rules': []
```
Comment (jmcarp) [inline website/addons/dropbox/routes.py]: Can `web_routes` be deleted?

File: website/addons/github/routes.py [hunk 13]
```diff
@@ -164,40 +143,10 @@
...
-                '/project/<pid>/node/<nid>/github/file/<path:path>/download/',
-            ],
-            'get',
-            views.crud.github_download_file,
-            notemplate,
-        ),
-    ],
-}
\ No newline at end of file
+    'rules': []
```
Comment (jmcarp) [inline website/addons/github/routes.py]: Can this also be deleted?

File: website/addons/github/static/githubFangornConfig.js [hunk 14]
```diff
@@ -169,17 +170,10 @@ function _fangornGithubTitle(item, col)  {
...
-                              path: item.data.path.substring(1),
-                              branch: item.data.branch
-                          },
-                          item.data.extra || {}
-                        )
-                    );
-                    window.location = item.data.nodeApiUrl + 'waterbutler/files/?' + params;
+                    var redir = new URI(item.data.nodeUrl);
+                    redir.segment('files').segment(item.data.provider).segment(item.data.path.substring(1));
+                    window.location = redir.toString() + '/?branch=' + item.data.branch;
```
Comment (jmcarp) [inline website/addons/github/static/githubFangornConfig.js]: How about `redir.search({branch: item.data.branch})`?

File: website/static/js/fileRevisions.js [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+    self.node = node;
+    self.file = file;
+    self.editable = editable;
+    self.urls = {
+        delete: waterbutler.buildDeleteUrl(file.path, file.provider, node.id),
+        download: waterbutler.buildDownloadUrl(file.path, file.provider, node.id),
+        revisions: waterbutler.buildRevisionsUrl(file.path, file.provider, node.id),
+    };
+    self.page = 0;
```
Comment (jmcarp) [inline website/static/js/fileRevisions.js]: Since we're not paginating revisions, can we delete all the associated code--references to `self.page`, `self.more`, etc?

File: website/templates/project/view_file.mako [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+                        {{ revision.extra.user.name }}
+                    <!-- /ko -->
+                    </td>
+                    <!-- /ko -->
+                    <td>
+                      <!-- ko if: revision.extra && revision.extra.downloads -->
+                      <span class="badge">{{ revision.extra.downloads }}</span>
+                      <!-- /ko -->
+                      <a class="btn btn-primary btn-sm" 
+                        href="{{ '?' + revision.versionIdentifier + '=' + revision.version + '&action=download' }}"
```
Comment (jmcarp) [inline website/templates/project/view_file.mako]: How about adding something like an `downloadOsf` field on the `Revision` model to take this logic out of the template?

File: website/templates/project/view_file.mako [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+
+        <!-- ko if: versioningSupported -->
+        <table class="table osfstorage-revision-table ">
+
+            <thead>
+                <tr>
+                    <th>Version</th>
+                    <th>Date</th>
+                    <!-- ko if: revisions()[0] && revisions()[0].extra && revisions()[0].extra.user -->
```
Comment (jmcarp) [inline website/templates/project/view_file.mako]: Can this be moved out of the template and documented?

File: website/templates/project/view_file.mako [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+        </a>
+        <!-- ko if: editable -->
+        <button class="btn btn-danger btn-md" data-bind="click: askDelete">
+          Delete <i class="icon-trash"></i>
+        </button>
+        <!-- /ko -->
+
+
+        <!-- ko if: versioningSupported -->
+        <table class="table osfstorage-revision-table ">
```
Comment (jmcarp) [inline website/templates/project/view_file.mako]: Can we drop the `osfstorage-revision-table` class? #copypaste

File: website/templates/project/view_file.mako [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+                    <th>User</th>
+                    <!-- /ko -->
+                    <th>Download</th>
+                </tr>
+            </thead>
+
+            <tbody data-bind="foreach: {data: revisions, as: 'revision'}">
+                <tr>
+                    <td>
+                      <a href="{{ '?' + revision.versionIdentifier + '=' + revision.version }}">
```
Comment (jmcarp) [inline website/templates/project/view_file.mako]: See other comments, prefer to keep logic out of templates.

File: website/templates/project/view_file.mako [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+                        {{ revision.version.substring(0, 8) }}
```
Comment (jmcarp) [inline website/templates/project/view_file.mako]: Can we move this `substring` to the view model?

File: website/static/js/filerenderer.js [hunk 21]
```diff
@@ -8,23 +8,37 @@ FileRenderer = {
...
             }
         });
+
+        fut.fail(self.handleRetry);
+    },
+
+    handleRetry: function() {
+        var self = this;
+        self.tries += 1;
+        if(self.tries > 10){
```
Comment (jmcarp) [inline website/static/js/filerenderer.js]: Can the number of tries be a named constant?

File: website/static/js/filerenderer.js [hunk 22]
```diff
@@ -8,23 +8,37 @@ FileRenderer = {
         this.url = url;
         this.element = $(selector);
         this.tries = 0;
-        this.refreshContent = window.setInterval(this.getCachedFromServer.bind(this), 1000);
+        // this.refreshContent = window.setInterval(this.getCachedFromServer.bind(this), 1000);
```
Comment (jmcarp) [inline website/static/js/filerenderer.js]: Can this be deleted?

File: website/static/js/fileRevisions.js [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+
+RevisionsViewModel.prototype.fetch = function() {
+    var self = this;
+    var request = $.getJSON(
+        self.urls.revisions,
+        {page: self.page}
+    );
+
+    request.done(function(response) {
+        // self.more(response.more);
```
Comment (jmcarp) [inline website/static/js/fileRevisions.js]: Delete this?

File: website/templates/project/view_file.mako [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+                        data-bind="click: revision.download">
+                        <i class="icon-download-alt"></i>
+                      </a>
+                    </td>
+                </tr>
+            </tbody>
+
+        </table>
+
+        <p data-bind="if: more">
```
Comment (jmcarp) [inline website/templates/project/view_file.mako]: We can delete this too.

File: framework/render/core.py [hunk 25]
```diff
@@ -19,3 +27,40 @@ def init_mfr(app):
...
+    # If none of the above go ahead and start
+    return False
+
+
+def save_to_file_or_error(download_url, dest_path):
+    with open(dest_path, 'wb') as temp_file:
+        response = requests.get(download_url, stream=True)
+        if response.ok:
+            for block in response.iter_content(1024):  # 1kb
+                if not block:
```
Comment (jmcarp) [inline framework/render/core.py]: This check isn't necessary--`iter_content` returns a generator that will exhaust when the response is empty.

File: website/addons/base/__init__.py [hunk 26]
```diff
@@ -166,17 +171,123 @@ class GuidFile(GuidStoredObject):
...
+
+    @property
+    def metadata_url(self):
+        url = self._base_butler_url
+        url.path.add('data')
+
+        return url.url
+
+    @property
+    def cache_path(self):
```
Comment (jmcarp) [inline website/addons/base/__init__.py]: To be more explicit, how about prefixing methods that are mfr-specific with mfr? e.g. `mfr_cache_path`

File: framework/render/core.py [hunk 27]
```diff
@@ -19,3 +27,38 @@ def init_mfr(app):
...
+
+
+def render_is_done_or_happening(cache_path, temp_path):
+    # if the cached file exists do nothing
+    if os.path.isfile(cache_path):
+        return True
+
+    if os.path.isfile(temp_path):
+        if time.time() - os.path.getmtime(temp_path) > settings.MFR_TIMEOUT:
+            # If the temp path has not been modified since the timeout
```
Comment (sloria) [inline framework/render/core.py]: :+1:

File: website/addons/base/views.py [hunk 28]
```diff
@@ -201,19 +207,117 @@ def create_waterbutler_log(payload, **kwargs):
...
+
+    if kwargs.get('path'):
+        path = kwargs['path']
+    elif kwargs.get('fid'):
+        path = kwargs['fid']
+
+    if kwargs.get('provider'):
+        provider = kwargs['provider']
+    elif 'osffiles' in request.path:
+        provider = 'osfstorage'
```
Comment (sloria) [inline website/addons/base/views.py]: There is no `else` block here, so `provider` may be undefined in the call toe `node.web_url_for`

File: website/addons/base/views.py (hunk 10 above)
Comment (sloria) [inline website/addons/base/views.py]: +1 @jmcarp

File: website/addons/base/views.py [hunk 29]
```diff
@@ -201,19 +207,117 @@ def create_waterbutler_log(payload, **kwargs):
...
+            **query_params
+        ),
+        code=httplib.MOVED_PERMANENTLY
+    )
+
+
+@must_be_valid_project
+@must_be_contributor_or_public
+def addon_view_or_download_file(auth, path, provider, **kwargs):
+    extras = request.args.to_dict()
```
Comment (sloria) [inline website/addons/base/views.py]: Can you please document the expected query parameters in a docstring? e.g..

```python
"""
:param-query str action: (Optional) Action to take. May be one of 'view' or 'download'.
"""
```

File: website/addons/github/model.py [hunk 30]
```diff
@@ -163,6 +170,20 @@ class AddonGitHubNodeSettings(AddonNodeSettingsBase):
 
     registration_data = fields.DictionaryField()
 
+    def find_or_create_file_guid(self, path):
+        try:
+            return GithubGuidFile.find_one(
+                Q('path', 'eq', path) &
+                Q('node', 'eq', self.owner)
+            ), False
+        except ModularOdmException:
```
Comment (sloria) [inline website/addons/github/model.py]: Better to catch `modularodm.exceptions.NoResultsFound` here.

Comment (jmcarp) [inline website/addons/github/model.py]: FWIW, this will have to be refactored anyway--I'll update #1601 and replace this logic once we're ready to get back to the issue of duplicate guids.
